### ISSUE TITLE
NLTK Naive Bayes Classifier train method runs out of memory for considerable train data size 

### ISSUE THREAD

**Issue by ghost** [nltk/nltk#2042]
I was using the Naive Bayes classifier for training a sample set.
The sample set has 11392 labels and 4109 unique features spread across the labels.

When the train method is called the memory runs out rapidly and Memory error is received.
I have done the following:
1. Moved the code to use 64 Bit Python interpreter
2. Used indices (int instead of string) for both features and labels
But still the memory is running out rapidly and the training never completes
On a 16 GB RAM system if the train method is called when Physical RAM under use is 2.9 GB, in a matter of minutes it shoots up to use the whole physical RAM and after that thrashing starts as disk usage starts for swapping pages. 

Any assistance would be highly appreciated

**Comment by alvations** [comment]
**TL;DR**

Take a look at https://www.kaggle.com/alvations/basic-nlp-with-nltk 

I suspect you might run out of memory too using scikit-learn too. 
Why do you have 11392 labels? Are you sure it's a task suitable for classification? How many data points do you have?

----

I think it's a little `off-topic` as it's not exactly a bug but yes you're right about naive bayes implementation in NLTK not being as efficient as other libraries. And a good chance to ponder over some NLTK related issues with regards to machine learning and speed.

Largely because the Naive Bayes in NLTK is used for 

 - from usage, teaching about linguistics features and supervised learning (in an education setting)
 - from code reading, learning how to implement the classifier in Python with as little speed-up and memory tricks as possible

Question is, should we wrap a faster implementation by default? Do people care about reading the code? Or just using the function? Should we reimplement more efficient classifier with memory and speed-up tricks?

**Comment by ghost** [comment]
Thanks @alvations for the reply and information.
The prototype being developed is for page classification for further document classification.
Currently there are 250-300 documents and each of them might have variations i.e. page content is different and/or number of pages are different.
Currently the learnset has around 7000 pages i.e. ~7000 labels
So in case the prototype is taken to product stage it should be able to handle more documents i.e. new ones coming in, and different variations i.e. for existing and new documents

Hence in light of the above the memory and speed requirements become important
